### PR TITLE
djl-bench 0.12.0 (new formula)

### PR DESCRIPTION
**Pull request by aksrajvanshi**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:16]: Is the `mv` necessary, or can we just do this instead?

```suggestion
    libexec.install Dir["*"]
    env = { APP_HOME: "${APP_HOME:-#{var}}" }
    env.merge!(Language::Java.overridable_java_home_env)
    (bin/"djl-bench").write_env_script libexec/"bin/benchmark", env
```

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:13]:
```suggestion
    prefix.install_metafiles
    libexec.install Dir["*"]
```
This makes sure things like license files, readmes, etc get installed into the prefix and not `libexec`.

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:16]:
```suggestion
    djl_env = { APP_HOME: "${APP_HOME:-#{var}}" }
    djl_env.merge!(Language::Java.overridable_java_home_env)
    (bin/"djl-bench").write_env_script "#{libexec}/bin/djl-bench", djl_env
```
The `Formula` class should have an `env` method, so let's try not to shadow that.

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:26]: Why is this necessary? Do you need write access to `APP_HOME`?

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:27]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

File: Formula/djl-bench.rb
Comment (aksrajvanshi) [reply Formula/djl-bench.rb:26]: Yes, I realized this isn't needed.

File: Formula/djl-bench.rb
Comment (aksrajvanshi) [reply Formula/djl-bench.rb:16]: So, we're running the package as `djl-bench ...`. I think I need the `mv` for that purpose. I'm not aware of a better way to run it. Do you have any suggestions ?

Comment (carlocab) [reply Formula/djl-bench.rb:16]: What about the suggestion in my previous comment? Does that not work?

Comment (aksrajvanshi) [reply Formula/djl-bench.rb:16]: No, unfortunately, that doesn't work as intended.

Comment (carlocab) [reply Formula/djl-bench.rb:16]: Hmm, yes, the script seems to check what name it was called as. That's fine then.

File: Formula/djl-bench.rb
Comment (aksrajvanshi) [reply Formula/djl-bench.rb]: Hmm, so it seems like I'm having trouble downloading the MXNet engine on the ARM based machine. 

The previous `djl-bench -h` command works with djl-bench when installed properly. Should I revert to that ?

Comment (carlocab) [reply Formula/djl-bench.rb]: Does `djl-bench` download `libmxnet.dylib` at runtime? We don't want that. We should build that from source instead. Also, isn't the test failure a sign that `djl-bench` isn't working?

Comment (aksrajvanshi) [reply Formula/djl-bench.rb]: Yes, the djl-bench downloads and uses `libmxnet.dylib` when it runs. Since, it's not able to download/use it here, it's not able to run. Its designed to download (and eventually cache) these dependencies at the first time of its run.

Comment (carlocab) [reply Formula/djl-bench.rb]: We should build `libmxnet.dylib` during `install` instead, so that it's not downloaded at runtime. We've removed formulae that do similar things from Homebrew/core before. Building `libmxnet.dylib` should also hopefully allow this to work on ARM.

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:4]: Is there no compressed tarball?

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:20]: Doesn't this just end up downloading pre-built dependencies? We should be building those from source. Is the download versioned and checksummed?

Comment (aksrajvanshi) [reply Formula/djl-bench.rb:20]: Yeah. My bad. I will look into building from source now. Which download are you asking about ?

File: Formula/djl-bench.rb
Comment (aksrajvanshi) [reply Formula/djl-bench.rb:4]: I don't think we have a further compressed version.

File: Formula/djl-bench.rb
Comment (carlocab) [reply Formula/djl-bench.rb]: The download of `libmxnet.dylib` and other pre-built dependencies it fetches.

File: Formula/djl-bench.rb
Comment (carlocab) [inline Formula/djl-bench.rb:23]: The macOS runners have no `apt-get`. `git` is already installed, so there's no need to install it. Let's skip the use of `ccache`. For the others, you can do
```ruby
depends_on "cmake"
depends_on "ninja"
depends_on "openblas"
depends_on "opencv"
```
You should add `=> :build` where appropriate (presumably this applies to `ninja` and `cmake`).

Comment (carlocab) [reply Formula/djl-bench.rb:23]: It's probably also simpler to make `mxnet` a separate formula so this one can just declare that as a dependency.

**Review by carlocab — CHANGES_REQUESTED** [review]
This still downloads pre-built dependencies as runtime, so I don't think we can merge this as-is.

**Comment by frankfliu** [comment]
> This still downloads pre-built dependencies as runtime, so I don't think we can merge this as-is.

Downloading those packages at runtime is a feature of this tool:
1. We based on user's input to decide which engine to use (This is user's choice, default use MXNet)
2. We based on os type and GPU type to download different version/flavor of the engine.

Let me provide a little bit context about tool:
1. djl-bench is designed to benchmark a wild range of deep learning mode, include MXNet, TensorFlow, PyTorch, OnnxRuntime, PaddlePaddle, XGBoost and more. With djl-bench, user can easily compare model performance between different deep learning engines.
2. For each above engine, they have different native libraries for mac and linux.
3. And for each GPU type (cuda version) also requires different distribution. (9.0, 9.2, 10.0, 10.1, 10.2, 11.0, 11.1, 11.3) each cuda package is around 1G.

The following options doesn't work for us:
1. It's not possible to package all dependencies into the .tar file (more 10G the total binary size)
2. And it's also doesn't make sense to build them from source. All those deep learning framework have super complicated dependencies, and require the system pre-install many dependencies and have a very strict gcc version. The build itself usually take 2-3 hours.
3. It's also doesn't make sense to make each of them as dependency. We don't own any of those engines. and the owner of these engines already released pip package. It doesn't make sense for them to release C++ binary as brew package either. Because the C++ binary cannot do anything itself.

@carlocab Is there a policy that a brew package should not download dependencies at runtime?

**Comment by carlocab** [comment]
> Is there a policy that a brew package should not download dependencies at runtime?

Downloading dependencies at runtime is, in my view, a circumvention of [this policy](https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae). If you have dependencies that we can't build, then it becomes hard to say that these dependencies are open source, and I suspect `djl-bench` just isn't suitable for packaging in Homebrew.

I'll let @Homebrew/core weigh in here in case someone else has a different opinion.

**Comment by claui** [comment]
@carlocab We’ve had a similar discussion about a homebrew-core formula which depended on Java bytecode downloaded from a third party.
If I remember correctly, the result (with which I personally disagree) was that we consider Java bytecode to be some kind of source/binary hybrid thing and dont’t treat it strictly like binary code so it therefore didn’t fall under the policy.

**Comment by carlocab** [comment]
@claui you edited your comment as I was typing mine 😄

Just as a quick summary for those who don't want to scroll up (I don't blame you; there's a lot to get through):

This (sometimes?) tries to download a `libmxnet.dylib` at runtime and fails rather miserably on ARM, because there doesn't appear to be a suitable pre-compiled one.

**Comment by frankfliu** [comment]
@carlocab Thanks for quick response. Our package itself is Java application. We download those engines on behalf of user.
Regarding ARM support, it's purely depends on each engine. PyTorch and MXNet will have official ARM release soon. Once they officially support ARM (We don't want to download nightly build), djl-bench will be able to benchmark models (PyTorch and MXNet) on ARM.

**Comment by SMillerDev** [comment]
>  If you have dependencies that we can't build, then it becomes hard to say that these dependencies are open source, and I suspect djl-bench just isn't suitable for packaging in Homebrew.

Yeah, especially because they are open source I don't think it should be downloading it's own native binaries.

**Comment by MikeMcQuaid** [comment]
> Downloading dependencies at runtime is, in my view, a circumvention of [this policy](https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae). 

I would actually be fine with this.

> This (sometimes?) tries to download a `libmxnet.dylib` at runtime and fails rather miserably on ARM, because there doesn't appear to be a suitable pre-compiled one.

I am not fine given this.

**Comment by frankfliu** [comment]
> > Downloading dependencies at runtime is, in my view, a circumvention of [this policy](https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae).
> 
> I would actually be fine with this.
> 
> > This (sometimes?) tries to download a `libmxnet.dylib` at runtime and fails rather miserably on ARM, because there doesn't appear to be a suitable pre-compiled one.
> 
> I am not fine given this.

We had added a check and make we display a proper message to user on ARM instead of throw an exception.

**Comment by danielnachun** [comment]
>All those deep learning framework have super complicated dependencies, and require the system pre-install many dependencies and have a very strict gcc version. The build itself usually take 2-3 hours.

You may also want to check if the pre-built dependencies work properly on Linux.  We have users who install Homebrew on Linux in VMs or HPC systems with old or minimal host libraries and these pre-built binaries won't know how to find the Homebrew versions of libraries either because they won't have the right RPATHs.
>It's also doesn't make sense to make each of them as dependency. We don't own any of those engines. and the owner of these engines already released pip package. It doesn't make sense for them to release C++ binary as brew package either. Because the C++ binary cannot do anything itself.

I don't think you should try to build those engines in this formula, or make them a dependency, but maybe we should eventually make formulae for those engines.  We already have `libtensorflow` and `libtorch`, and anything which has a `pip` package could easily be made into a formula here.  It's worth remembering that a `pip` package with native code will either have to be compiled from source by the user, or may ship a broken native binary, whereas with a Homebrew formula we can provide the user a binary that we know will work.  

If it does not already do so, it might be a good idea for `djl-bench` to allow the user to use an engine which has already been downloaded and installed elsewhere.  That would give them the flexibility to install engines using Homebrew (or Anaconda) if the binaries you try to install for them don't work.   

I don't want to hold up approval of this formula, I just wanted to point out some of the challenges in providing native binaries on Linux to users.

**Comment by frankfliu** [comment]
djl-bench do support Linux (centos 7+, ubuntu 16.04 +, Amazon Linux 2).
We provide deb installation package: https://publish.djl.ai/djl-bench/0.12.0/djl-bench_0.12.0-1_all.deb

We are not able to depends on libtensorflow or libtorch:
1. We support so many flavor of libtorch: cpu:osx-x86_64, cu92, cu101, cu102, cu111, cu113, cpu-precxx11, and we are planing to support aarch64 by the end of this year. The libtorch fomula seems only support intel based cpu on mac.
2. We cannot use libtensorflow either, Google released sightly different libtensorflow.so for Java, we can't use pip version of libtensorflow.so 

And we actually allow user to use their own binary by setting environment variable, for example:
```
export PYTORCH_LIBRARY_PATH=/Library/Python/3.9/site-packages/torch/
```

Downloading the binary dependency is just trying to provide the best user experience. It's one the feature of this tool.

**Comment by SMillerDev** [comment]
> The libtorch fomula seems only support intel based cpu on mac.

That's because macs only ever have an intel or ARM CPU, and we can't build libtorch for ARM yet. Someone using another CPU won't be able to install djl-bench with homebrew anyway.

**Comment by danielnachun** [comment]
> djl-bench do support Linux (centos 7+, ubuntu 16.04 +, Amazon Linux 2).
> We provide deb installation package: https://publish.djl.ai/djl-bench/0.12.0/djl-bench_0.12.0-1_all.deb

I should have added that our goal in general is to support all Linux distros (in theory the only requirement is an Intel CPU and Linux kernel 2.6+), and to work without sudo access.  For that reason we try to avoid scenarios where a system package manager needs to be used to make things work, as this requires sudo access.

> We are not able to depends on libtensorflow or libtorch:
> 
> 1. We support so many flavor of libtorch: cpu:osx-x86_64, cu92, cu101, cu102, cu111, cu113, cpu-precxx11, and we are planing to support aarch64 by the end of this year. The libtorch fomula seems only support intel based cpu on mac.
> 2. We cannot use libtensorflow either, Google released sightly different libtensorflow.so for Java, we can't use pip version of libtensorflow.so

That is something we can probably improve in Homebrew, though I don't think it's needed right now.

> And we actually allow user to use their own binary by setting environment variable, for example:
> 
> ```
> export PYTORCH_LIBRARY_PATH=/Library/Python/3.9/site-packages/torch/
> ```
> 
> Downloading the binary dependency is just trying to provide the best user experience. It's one the feature of this tool.

Since you provide the user flexibility to install the engines using other systems (even if Homebrew doesn't currently provide formulae that would work for this), I think this is fine for Linux users.  I would be more concerned if they could only use the binaries downloaded by `djl-bench` given that they could be broken, but since that is optional behavior it seems okay to me.

**Comment by MikeMcQuaid** [comment]
> We had added a check and display a proper message to user on ARM instead of throw an exception.

Thanks for this!

> I should have added that our goal in general is to support all Linux distros

To clarify: "work on" rather than "support".

---

> * We support so many flavor of libtorch: cpu:osx-x86_64, cu92, cu101, cu102, cu111, cu113, cpu-precxx11, and we are planing to support aarch64 by the end of this year. The libtorch fomula seems only support intel based cpu on mac.

How are these used? Are you cross-compiling for these architectures?

**Comment by frankfliu** [comment]
@MikeMcQuaid 
1. We directly load libtorch.so if PYTORCH_LIBRARY_PATH is specified
2. We detect os.name, os.arch, and use JNA to detect cuda version and cuda arch, and based on those information we download pre-compiled libtorch.so file. For most of the framework, we don't compile them ourselves, we got the binary from each engine's official release.

**Comment by MikeMcQuaid** [comment]
@frankfliu sorry, what I'm asking for: is the tool a cross-compiler or similar? If not, why do you care about all those arches?

**Comment by frankfliu** [comment]
> @frankfliu sorry, what I'm asking for: is the tool a cross-compiler or similar? If not, why do you care about all those arches?

No, the tools is not a cross-compiler, this a benchmarking tool that allows user understand the machine model's performance in different Engine/platform. For example, I have an image classification model, and my latency budget is 15 ms and throughput is 300 image/s. I want to know with how many CPUs + GPUs + use which engine can should I deploy my model. After run all the combination, I may conclude that PyTorch + 4 CUDA VT100 GPUs + 96 CPUs threads can meet my requirement.

**Comment by MikeMcQuaid** [comment]
> No, the tools is not a cross-compiler, this a benchmarking tool that allows user understand the machine model's performance in different Engine/platform.

Ok, so why is "The libtorch fomula seems only support intel based cpu on Mac." a problem if I'm installing it on an Intel CPU on macOS?

**Comment by vineethnara** [comment]
djl user here and I thought I could add perspective here if you do not mind. I use Linuxbrew with some of my Linux systems and the reason djl getting the right version of torch is better is because different systems may have different versions of CUDA installed based on user choice and the NVIDIA GPU available.

You are right that for Intel Macs this is not an issue to use the formula because Macs don't have CUDA support at the moment. Torch is not yet natively available for Apple Silicon Macs so that is not an immediate concern either.

But for Linux systems which may have any version of CUDA the right version of torch must be installed with CUDA support, which the formula doesn't provide. So it is better that djl gets its own libtorch based on the CUDA version if CUDA is available or the CPU version of lib torch if I am benchmarking on CPU and not only GPU.

**Comment by aksrajvanshi** [comment]
Do we have a final consensus on downloading native pre-built libraries for djl-bench brew package ?

**Comment by MikeMcQuaid** [comment]
@akshayakrsh The current consensus seems to be "it's not OK". I'm asking more questions so I may be able to change my mind, though.

**Pull request closed without merging**
